File: csharp/circular-buffer/CircularBuffer.cs
using System;

public class CircularBuffer<T>
{
    private T[] _data;
    private int _readPos;
    private int _writePos;

    public CircularBuffer(int capacity)
    {
        if (capacity < 1) { throw new ArgumentException(); }
        _data = new T[capacity];

        _readPos = -1;
        _writePos = 0;
    }

    public T Read()
    {
        if (_readPos == -1) {
            throw new InvalidOperationException();
        }

        T value = _data[_readPos];
        _readPos++;

        return value;
    }

    public void Write(T value)
    {
        // TODO Is this the best way to manage this
        if (_readPos == -1) { _readPos = 0; }

        _data[_writePos] = value;
        _writePos++;
        _writePos %= _data.Length;
    }

    public void Overwrite(T value)
    {
        throw new NotImplementedException("You need to implement this function.");
    }

    public void Clear()
    {
        int capacity = _data.Length;
        _data = new T[capacity];
    }
}

File: java/circular-buffer/src/main/java/CircularBuffer.java
public class CircularBuffer<T> {

    private static final String MSG_BUFFER_FULL = "Tried to write to full buffer";
    private static final String MSG_BUFFER_EMPTY = "Tried to read from empty buffer";

    private int size;
    private T buffer[];
    private int readPos;
    private int writePos;

    @SuppressWarnings("unchecked")
    public CircularBuffer(int size) {

        readPos = 0;
        writePos = 0;
        this.size = size;
        buffer = (T[])new Object[size];
    }

    @SuppressWarnings("unchecked")
    public void clear() {
        buffer = (T[])new Object[size];
    }

    private boolean canWrite() {
        return buffer[writePos] == null;
    }

    private void forceWrite(T value) {
            buffer[writePos] = value;
            writePos = (writePos + 1) % buffer.length;
    }

    public void overwrite(T value) {

        // if space left in buffer, use that
        if (canWrite()) {
            forceWrite(value);
        } else {
            // otherwise use oldest data
            buffer[readPos] = value;
            readPos = (readPos + 1) % buffer.length;
        }
    }

    public T read() throws BufferIOException {

        T value = buffer[readPos];

        if (value == null) {
            throw new BufferIOException(MSG_BUFFER_EMPTY);
        }

        buffer[readPos] = null;
        readPos = (readPos + 1) % buffer.length;

        return value;
    }

    public void write(T value) throws BufferIOException {

        if (canWrite()) {
            forceWrite(value);
        } else {
            throw new BufferIOException(MSG_BUFFER_FULL);
        }
    }
}